WP3/D3.8/report.tex

```diff
@@ -185,13 +185,94 @@ \section{Continuous integration achievements of OpenDreamKit projects}\label{pro
+advisory tool for developers but to catch problems and incompatibilities of
+Sage or its dependencies, in particular on more obscure platforms that
+dependent projects do often not test for. Due to the vast range of platforms
+covered, the Buildbot can not realistically be turned into an advisory CI
+system as it is not conceivable to provision and administer the number of
+machines that would be necessary to run Sage's test suite at every stage of a
+test that gets proposed by a developer.
+
+Therefore, the Sage community at some point developed its own patchbot system
+which runs Sage's test suite at every stage of a proposed change. Anybody can
+run a patchbot on their private computer, typically a Linux system. These
+patchbots will then look for proposed changes to Sage, run Sage's test suite,
+and report findings back to a central server. It is worth noting that the
+patchbot system predates most of the currently popular CI systems.
+
+The patchbot system comes with a number of issues: It is not nearly as polished
+as the professional offerings which means that its workings can often be
+confusing and that it also requires continuous maintenance and development from
+the SageMath community and by those that run the patchbots on their private
+machines. The fact that it is running on private machines means that due to
+load, connectivity issues, and other temporary issues on these machines, the
+patchbots are very frequently reporting false positives. This noise can be very
+frustrating as it makes the patchbot system unreliable and in particular harder
+to understand for new contributors. Finally, due to limitations of the patchbot
+architecture, not all types of changes can be tested, in particular it can not
+test changes that involve dependent software packages. This can be frustrating
+as testing such changes is often the most time consuming task for developers
+and it would be particularly nice to give such tasks away to a CI system.
+
+\subsubsection{Modern CI with GitLab CI and CircleCI}
+To fix these issues we were looking at best pratices that would give the
+SageMath community fast, reproducible, reliable CI with minimal maintenance and
+continuous development necessary to keep the CI working for everybody.  We
+settled for GitLab CI, a modern open-source CI system, and CircleCI, a popular
+closed-source system. Which one to use is up to each developer. For simplicity
+we focus on GitLab CI here. GitLab CI typically uses docker containers as its
+backend (but it can also be based on many other technologies.) These containers
+are essentially virtual machines that can be hosted in a central location but
+also on private computers. Unlike in the patchbot systems, docker containers
+are standardized and isolated so that there is usually no issue of false
+positives. % note that I do not distinguish between docker containers and
+% images; I think it's more confusing to the reader to make such a distinction.
```

Comment:
I agree, though later on I might say something more about this distinction.  @jdemeyer raised an interesting point that he would like to be able to test Cysignals on more platforms.  And in the case of Cysignals in particular I can imagine cases where one would like to be able to test it against different kernels, which is not possible with containers.

Reply:
True.